### ISSUE TITLE
Feature: cached views (for toggling visibility)

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#4821]
## 🚀 Feature
It's nice that we don't unnecessarily render layers that are not visible, but it's a shame that toggling visibility off and back on is now a rather expensive operation (taking about 5 seconds on my computer).  I very frequently want to toggle the visibility of a labels layer to evaluate how i'm doing, but it's kinda infeasible for 3D data at the moment. 

### Pitch

maybe we could have a cache that somehow uses the last used visible view if it hasn't been invalidated in the meantime?

**Comment by alisterburt** [comment]
Love the proposal - I don't think the way you've proposed it will work exactly (don't think we can cache intermediate framebuffers from separate visuals) but simply toggling visibility and debouncing the offloading of data from VRAM (long timescale, 10s or so?) could be a great solution

**Comment by alisterburt** [comment]
cc @brisvag

**Comment by brisvag** [comment]
Yes! This always annoys me, but I never got around to do anything about it.

I'm not sure if there's a way to keep the data on vram on the vispy side when `visible` is set to false... But are we sure that's the bottleneck? Isn't it the call to `Layer.refresh()` which then does:

```py
        if self.visible:
            self.set_view_slice()
            self.events.set_data()  # refresh is called in _update_dims which means that extent cache is invalidated. Then, base on this event extent cache in layerlist is invalidated.
            self._update_thumbnail()
            self._set_highlight(force=True)
```

`self.events.set_data()` being probably the major offender here?

**Comment by alisterburt** [comment]
not sure, would need to check the details - once all set in vispy interacting is super fast so I assumed the slow thing was a transfer back to vispy, aren't we saying the same thing? (i.e. shouldn't need to call set_data because the data hasn't changed)

**Comment by brisvag** [comment]
Sorry, I meant "set view slice" :)

**Comment by psobolewskiPhD** [comment]
> it's a shame that toggling visibility off and back on is now a rather expensive operation (taking about 5 seconds on my computer).

Even with a very large layer (a whole slide image or multi-Gb 3D image), toggling visibility is basically instant for me.
Has the machinery been optimized/replaced since this issue was made?

**Comment by brisvag** [comment]
The whole refresh machinery has been touched quite a bit, also recently, which might have made this operation a lot cheaper over time.

We do our best to only refresh what's necessary (so we do cache things like extent), but we're still not caching in the sense @tlambert03 probably meant (i.e: no cached data on the GPU, so the CPU->GPU transfer operation still needs to happen).

**Comment by tlambert03** [comment]
Yep that's what I meant.  But feel free to close this if you need cleanup and that seems unlikely

**Comment by brisvag** [comment]
I'd leave open. It actually shouldn't be too hard, we are pretty close to it.
Currently, data *already* stays on the GPU when invisible. However, as soon as we make visible again, we issue a `refresh` because things might have changed while invisible, and we ignore all those changes to avoid losing performance due to invisible layers.

We just need to be smarter about how we cache the current "view state" and only send new data when stricly necessary. This is probably a lot easier to do once we finish migrating all layers to the `async` project (which we are already planning to do soon with https://github.com/napari/docs/pull/730), because then we just need to compare a slice request with the previous one and only resend data if critical things differ.

On the other hand, this can already be done if someone wanted to, by just setting a flag such as `Layer.needs_data_update` when necessary, and on only if that's set firing the `set_view_slice` machinery.
